### PR TITLE
[Proposal] Support loading from safetensors if file is present.

### PR DESCRIPTION
**Pull request by Narsil**
Current proposed behavior mimics transformers.

- Do not enforce `safetensors`.
- If `safetensors` is present AND the file is present (on the remote model or locally).
  the load from it preferrably.

### CODE REVIEW

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot for the PR @Narsil ! The code looks good, and I'm in favor of supporting `safetensors`. Is there any example repo with `safetensors` weights which we could try ?

Also, don't we need any changes to `save_pretrained` to save weights in this format ?

**Comment by Narsil** [comment]
> Also, don't we need any changes to save_pretrained to save weights in this format ?

Currently, I didn't do any change.

`pipeline.save_pretrained` will call all modules methods on its own which will make it harder to use safetensors.
`model.save_pretrained` does support `save_function` which makes it easy to support safetensors.

However, I don't think you can override `save_function` within the `pipeline.save_pretrained`. 

Overall I would be more conservative with saving weights as an opt-in rather than a default to keep things more gradual in changes (since saving in safetensors meaning users will have to have it to load the weights). Actually we could load the weights even without safetensors because the loading is super simple https://gist.github.com/Narsil/3edeec2669a5e94e4707aa0f901d2282
I'm just being conservative in changing patterns.

> Thanks a lot for the PR @Narsil ! The code looks good, and I'm in favor of supporting safetensors. Is there any example repo with safetensors weights which we could try ?
Not yet, I'm in the process of updating https://github.com/huggingface/safetensors/blob/main/bindings/python/convert.py to support ANY framework by default (just convert ALL pytorch weights of the repo and just change the extension).

**Comment by patrickvonplaten** [comment]
Design also looks good to me! Thanks a lot for working on it @Narsil !

**Comment by Narsil** [comment]
@patil-suraj 

https://huggingface.co/Narsil/stable-diffusion-v1-4/
https://huggingface.co/Narsil/tiny-stable-diffusion-torch

Are now converted.

Simply 
```
git clone https://github.com/huggingface/safetensors
cd bindings/python
python convert.py MODEL_ID
```

Or simply this: https://huggingface.co/spaces/safetensors/convert

**Comment by Narsil** [comment]
Should I create a test for this ? (Could have a tiny random pipeline with only safetensors weights to test, but I would need to modify the CI somewhere to install it.)

**Comment by patrickvonplaten** [comment]
> Should I create a test for this ? (Could have a tiny random pipeline with only safetensors weights to test, but I would need to modify the CI somewhere to install it.)

Yes a test would be great! Could you maybe just `safetensors` here: https://github.com/huggingface/diffusers/blob/44e56de9aaaa103ad11ca2953dc86ba6f64ba5d4/setup.py#L189 ?

**Comment by patrickvonplaten** [comment]
It should then be used automatically be the Github Runner (cc @anton-l )

**Comment by patil-suraj** [comment]
The loading time is crazy fast, I'm getting ~1.3 sec 🔥

**Comment by Narsil** [comment]
> The loading time is crazy fast, I'm getting ~1.3 sec on CPU fire

should be lower than that, I think I know why, I found a bug creating the test, I'm fixing everything right now.

**Comment by Narsil** [comment]
Okay I  think you should check again the logic as I'm not confident about this change.

`snapshot_download` uses `allow` and `ignore` lists to get the snapshot in a ready to go fashion.

Since both `.safetensors` files and `.bin` files were accepted before it downloaded both versions and everything still worked. However, it was loading the Pytorch files for `transformers` (see https://github.com/huggingface/safetensors/pull/105).

This modification does the fix, but adds an extra network call to fetch the model_info to filter out the files *before* snapshot_download. This is the smallest change I could think of, but maybe we can do even better (and remove the `*.msgpack` oddity at the same time.

However this feels like a larger change, which would require a separate PR.

The idea would be to remove all network calls when using `_OFFLINE` and make sure the CPU load happens in a few ms (currently sitting at 1.6s for me and my lousy network, and with offline options it goes down only to 1.2, 1.3s which is not normal)

**Comment by Narsil** [comment]
Also if that's interesting I could open another PR to load the pipeline directly on CUDA, this will make a difference with safetensors (and the appropriate flag).

File: src/diffusers/modeling_utils.py [hunk 1]
```diff
@@ -689,3 +642,101 @@ def unwrap_model(model: torch.nn.Module) -> torch.nn.Module:
...
+
+
+def get_model_file(pretrained_model_name_or_path, **kwargs) -> Tuple[str, bool]:
+    if is_safetensors_available():
+        try:
+            model_file = _get_model_file(
+                pretrained_model_name_or_path, weights_name=SAFETENSORS_WEIGHTS_NAME, **kwargs
+            )
+            return model_file, False
+        except:
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_utils.py:654]: Could we catch with a more explicit "File doesn't exist" error message here?

File: src/diffusers/modeling_utils.py [hunk 2]
```diff
@@ -689,3 +642,101 @@ def unwrap_model(model: torch.nn.Module) -> torch.nn.Module:
         return unwrap_model(model.module)
     else:
         return model
+
+
+def get_model_file(pretrained_model_name_or_path, **kwargs) -> Tuple[str, bool]:
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_utils.py:647]: The functions become a bit too nested for me here and hard to follow - could we maybe just remove this function and only keep `_get_model_file`? I don't think those 8 lines deserve a new function here, I'd prefer to just call `_get_model_file` directly above.

File: src/diffusers/modeling_utils.py [hunk 3]
```diff
@@ -689,3 +642,101 @@ def unwrap_model(model: torch.nn.Module) -> torch.nn.Module:
...
+                pretrained_model_name_or_path, weights_name=SAFETENSORS_WEIGHTS_NAME, **kwargs
+            )
+            return model_file, False
+        except:
+            pass
+    model_file = _get_model_file(pretrained_model_name_or_path, weights_name=WEIGHTS_NAME, **kwargs)
+    return model_file, True
+
+
+def _get_model_file(
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_utils.py:660]: nice!

File: src/diffusers/modeling_utils.py [hunk 4]
```diff
@@ -466,7 +419,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
             # if device_map is Non,e load the state dict on move the params from meta device to the cpu
             if device_map is None:
                 param_device = "cpu"
-                state_dict = load_state_dict(model_file)
+                state_dict = load_state_dict(model_file, is_pytorch)
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_utils.py:422]: Actually could we not use a flag here at all? And instead decide in `load_state_dict` depending on the model file name (it has to end with either `SAFETENSORS_WEIGHTS_NAME` or `WEIGHTS_NAME` whether it's `is_pytorch` or not. It's a bit confusing to me to pass around a `is_pytorch` flag here

File: src/diffusers/modeling_utils.py [hunk 5]
```diff
@@ -82,12 +87,16 @@ def find_tensor_attributes(module: torch.nn.Module) -> List[Tuple[str, Tensor]]:
         return first_tuple[1].dtype
 
 
-def load_state_dict(checkpoint_file: Union[str, os.PathLike]):
+def load_state_dict(checkpoint_file: Union[str, os.PathLike], is_pytorch: bool):
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_utils.py:90]:
```suggestion
def load_state_dict(checkpoint_file: Union[str, os.PathLike]):
```

File: src/diffusers/modeling_utils.py [hunk 6]
```diff
@@ -82,12 +87,16 @@ def find_tensor_attributes(module: torch.nn.Module) -> List[Tuple[str, Tensor]]:
...
     """
     Reads a PyTorch checkpoint file, returning properly formatted errors if they arise.
     """
     try:
-        return torch.load(checkpoint_file, map_location="cpu")
+        if is_pytorch:
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_utils.py:95]:
```suggestion
        if Path(checkpoint_file).name == WEIGHTS_NAME:
```

File: src/diffusers/modeling_utils.py [hunk 7]
```diff
@@ -82,12 +87,16 @@ def find_tensor_attributes(module: torch.nn.Module) -> List[Tuple[str, Tensor]]:
...
+            return torch.load(checkpoint_file, map_location="cpu")
+        else:
+            return load_file(checkpoint_file, device="cpu")
```
Comment (patrickvonplaten) [inline src/diffusers/modeling_utils.py:98]: just for readability could we maybe do the following:
```suggestion
            return safetensors.torch.load_file(checkpoint_file, device="cpu")
```

then the reader directly knows the loading comes from safetensors?

File: src/diffusers/modeling_utils.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/modeling_utils.py:654]:
```suggestion
        except EnvironmentError:
```

It can only be an `EnvironmentError` no?

**Comment by patrickvonplaten** [comment]
PR looks very nice!

I left some feedback regarding readability - always a bit worried to introduce too many nested functions and flags.

Think the download as implemented for the models now works really well. However, we're actually having a small problem with the snapshot download method I think. I'll open a Slack channel to discuss there as I think it'll be a bit tricky.

File: src/diffusers/pipeline_utils.py [hunk 8]
```diff
@@ -463,6 +463,28 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                 user_agent["custom_pipeline"] = custom_pipeline
             user_agent = http_user_agent(user_agent)
 
+            info = model_info(
```
Comment (patrickvonplaten) [inline src/diffusers/pipeline_utils.py:466]: can we wrap all of this into a function, called `is_compatible_with_safetensors()`? 

And then just do something like:

```python
if is_safetensors_available() and is_compatible_with_safetensors():
    ignore_patterns.append("*.bin")
```

**Comment by patrickvonplaten** [comment]
PR looks very nice!

I left some feedback in modeling_utils regarding readability - always a bit worried to introduce too many nested functions and flags.

Overall I think this is a really nice implementation! In pipeline_utils it would just be nice to also only do the extra call to the Hub if safetensors are installed and maybe move the whole logic into one simple sounding "is pipe compatible with safetensors" function

File: src/diffusers/modeling_utils.py (hunk 6 above)
Comment (Narsil) [reply src/diffusers/modeling_utils.py:95]: I feel this is wrong, we already know at this point what kind of file it is so we don't have to guess.

I agree the `is_pytorch` is slightly odd but I don't think doing things based on filename is great. 
I'll try to figure something out that's better.

File: src/diffusers/modeling_utils.py (hunk 7 above)
Comment (Narsil) [reply src/diffusers/modeling_utils.py:98]: Sure.

File: src/diffusers/modeling_utils.py (hunk 4 above)
Comment (Narsil) [reply src/diffusers/modeling_utils.py:422]: Again I think the behavior should not on the filename, but on what you did previously.

It's bikeshedding at this point, I'll try to find something even more elegant.

File: src/diffusers/modeling_utils.py (hunk 2 above)
Comment (Narsil) [reply src/diffusers/modeling_utils.py:647]: Happy to remove it, but I think the calling function will get messier.

File: src/diffusers/pipeline_utils.py (hunk 8 above)
Comment (Narsil) [reply src/diffusers/pipeline_utils.py:466]: We can.

I'd rather leave `model_info` on its own though.

It's a network call, and function names like `is_compatible` seem innocuous and fast (which doing network operations is not :))

File: src/diffusers/modeling_utils.py (hunk 6 above)
Comment (Narsil) [reply src/diffusers/modeling_utils.py]: Couldn't find anything satisfactory to me so I went your way.

**Comment by Narsil** [comment]
> PR looks very nice!
> 
> I left some feedback in modeling_utils regarding readability - always a bit worried to introduce too many nested functions and flags.

I don't necessarily agree but I went your way, I feel it's more style preferences so happy to oblige.

> 
> n pipeline_utils it would just be nice to also only do the extra call to the Hub if safetensors are installed and maybe move the whole logic into one simple sounding "is pipe compatible with safetensors" function

Hmm we should probably just own that call is what I'm suggesting since it's exactly what snapshot download does:
https://github.com/huggingface/huggingface_hub/blob/22c1431c598dca24e74efa2bed803df4a58fd862/src/huggingface_hub/_snapshot_download.py#L173

This calls for a larger PR though. 
Also using `safetensors` files could also be leveraged in Flax codebase (yet another PR)

`snapshot_download` assumes you're knowing which files you need uniquely based on where you are in the codebase (Flax vs PT). Assumption which this PR is breaking since now it's (what's on the servers + what's in the environment combination).

File: src/diffusers/pipeline_utils.py [hunk 9]
```diff
@@ -473,6 +491,15 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                 user_agent["custom_pipeline"] = custom_pipeline
             user_agent = http_user_agent(user_agent)
 
+            if is_safetensors_available():
```
Comment (patrickvonplaten) [inline src/diffusers/pipeline_utils.py:494]: cool that works for me - agree with your point that "calling the hub" is too hidden when fully put in `is_safetensors_compatible(...)`

File: src/diffusers/pipeline_utils.py [hunk 10]
```diff
@@ -117,6 +118,23 @@ class AudioPipelineOutput(BaseOutput):
...
+    is_safetensors_compatible = True
+    for pt_filename in pt_filenames:
+        prefix, raw = os.path.split(pt_filename)
+        if raw == "pytorch_model.bin":
+            # transformers specific
+            sf_filename = os.path.join(prefix, "model.safetensors")
+        else:
+            sf_filename = pt_filename[: -len(".bin")] + ".safetensors"
+        if sf_filename not in filenames:
+            print("{sf_filename} not found")
```
Comment (patrickvonplaten) [inline src/diffusers/pipeline_utils.py:133]:
```suggestion
            logger.warning("{sf_filename} not found")
```

File: src/diffusers/utils/import_utils.py [hunk 11]
```diff
@@ -42,6 +42,7 @@
 USE_TF = os.environ.get("USE_TF", "AUTO").upper()
```
Comment (patrickvonplaten) [inline src/diffusers/utils/import_utils.py:42]: Actually we should remove this -> we don't have TF in `diffusers` yet

File: src/diffusers/utils/import_utils.py [hunk 12]
```diff
@@ -55,7 +56,7 @@
         except importlib_metadata.PackageNotFoundError:
             _torch_available = False
 else:
-    logger.info("Disabling PyTorch because USE_TF is set")
+    logger.info("Disabling PyTorch because USE_TORCH is set")
```
Comment (patrickvonplaten) [inline src/diffusers/utils/import_utils.py:59]:
```suggestion
    logger.info("Disabling PyTorch because USE_TORCH is not set")
```

File: src/diffusers/utils/import_utils.py (hunk 11 above)
Comment (patrickvonplaten) [reply src/diffusers/utils/import_utils.py:42]: happy to do in a follow up PR

File: src/diffusers/utils/import_utils.py [hunk 13]
```diff
@@ -109,6 +110,17 @@
...
+if USE_SAFETENSORS in ENV_VARS_TRUE_AND_AUTO_VALUES:
+    _safetensors_available = importlib.util.find_spec("safetensors") is not None
+    if _safetensors_available:
+        try:
+            _safetensors_version = importlib_metadata.version("safetensors")
+            logger.info(f"Safetensors version {_safetensors_version} available.")
+        except importlib_metadata.PackageNotFoundError:
+            _safetensors_available = False
+else:
+    logger.info("Disabling Safetensors because USE_TF is set")
```
Comment (patrickvonplaten) [inline src/diffusers/utils/import_utils.py:122]:
```suggestion
    logger.info("Disabling Safetensors because USE_SAFETENSORS is not set")
```

**Comment by patrickvonplaten** [comment]
> I don't necessarily agree but I went your way, I feel it's more style preferences so happy to oblige.

Thanks for making the changes - fully agree that it's more style preference! 

Left some comments about `USE_TF` - in general we can remove all of those => we don't have TF in `diffusers` yet (if it's easy to do in `import_utils.py`, maybe we could do it in this PR - otherwise happy to do it myself in a follow-up PR

Happy to merge from my side though!

File: src/diffusers/pipeline_utils.py [hunk 14]
```diff
@@ -117,6 +118,23 @@ class AudioPipelineOutput(BaseOutput):
     audios: np.ndarray
 
 
+def is_safetensors_compatible(info) -> bool:
+    filenames = set(sibling.rfilename for sibling in info.siblings)
+    pt_filenames = set(filename for filename in filenames if filename.endswith(".bin"))
+    is_safetensors_compatible = True
```
Comment (patrickvonplaten) [inline src/diffusers/pipeline_utils.py:124]: Just thinking of an edge case. If we have a folder that has no PyTorch weights but only Flax weights and we want to load it into `DiffusionPipeline` with `from_pt=True` then `is_safetensors_compatible` would return true even though there are no `.safetensors` in the repo, but just because `pt_filenames` is an empty list. It's not really a problem right now because if there are no PyTorch weights we can safety add `*.bin` to the ignore list, but it might make the function a bit "incorrect". 

Could it maybe be slightly cleaner here to do:
```suggestion
    is_safetensors_compatible = any(file.endswith(".safetensors") for file in filenames)
```

This way we say the repo is compatible with safetensors as long as at least one safetensors file exists

**Review by patrickvonplaten — APPROVED** [review]
Nice PR - thanks a lot!

File: src/diffusers/pipeline_utils.py (hunk 14 above)
Comment (Narsil) [reply src/diffusers/pipeline_utils.py]: This is would also fail because `from_pt=True` would still fail since there are no pytorch models.

But the error would be much better with your fix. Good catch !
